File: android/src/main/kotlin/io/sentient/android/chat/composer/MicCorner.kt
// ---------------------------------------------------------------------------
// MicCorner — hold-to-talk / drag-to-lock mic control welded onto the composer
// card's top-right edge (half overhanging). Ports the webui MicCorner
// (gateway/webui/src/components/dock/mic-corner.tsx) — squircle shape, "ember
// trail" colorway, springy motion.
//
// Modes: IDLE | HOLD | LOCKED — mic on ⇔ HOLD or LOCKED.
//   press          → HOLD, mic starts immediately
//   drag LEFT      → button tracks the finger raw along the travel rail (snap)
//   release ≥ 40%  → LOCKED (snap to the far end, haptic); < 40% → IDLE
//                    (spring back, mic stops)
//   from LOCKED    → press again (base = travel), drag back right; release
//                    ≤ 50% of travel → IDLE (mic stops, haptic); else LOCKED
//
// The pure FSM (clampDrag / resolveRelease / isArmed) lives in
// MicCornerGesture.kt; the ember visuals live in MicCornerVisuals.kt.
//
// External sync: when [micActive] flips true→false while the control is not
// IDLE and not being dragged (disconnect, teardown, failed mic start), the
// control resets to IDLE WITHOUT emitting an intent. Only the true→false
// edge is observed so it never races the optimistic hold that begins before
// micActive propagates.
//
// Permission: [ensureMicPermission] gates the IDLE press. When it returns
// false (system prompt launched) the gesture is abandoned — no HOLD, no mic
// start. Stopping from LOCKED never checks permission.
//
// The wrap consumes its own pointer events (down + moves), so the composer's
// vertical swipe-dismiss gesture never fights the horizontal drag.
//
// testTags: "chat-mic" (the button, in MicCornerVisuals); the wrap flips
// "mic-corner" → "mic-corner-locked" while LOCKED for Maestro observability.
// ---------------------------------------------------------------------------
package io.sentient.android.chat.composer

import androidx.compose.foundation.gestures.awaitEachGesture
import androidx.compose.foundation.gestures.awaitFirstDown
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.setValue
import androidx.compose.animation.core.Animatable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.voice.talk.TalkMode
import kotlin.math.roundToInt
import kotlinx.coroutines.launch

private val log = createLogger("android", "mic-corner")

/** Wrap width — button + travel rail (webui: 96px wrap / 34px button, scaled). */
internal val MIC_CORNER_WRAP_WIDTH = 104.dp

/** Control button size — matches the composer row's [BUTTON_SIZE]. */
internal val MIC_CORNER_BUTTON = BUTTON_SIZE

/** Horizontal travel of the drag rail (wrap − button ≈ 66dp). */
internal val MIC_CORNER_TRAVEL = MIC_CORNER_WRAP_WIDTH - MIC_CORNER_BUTTON

@Composable
internal fun MicCorner(
    talkMode: TalkMode,
    ensureMicPermission: () -> Boolean,
    onPress: () -> Unit,
    onRelease: () -> Unit,
    onLock: () -> Unit,
    onStopContinuous: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val density = LocalDensity.current
    val travelPx = with(density) { MIC_CORNER_TRAVEL.toPx() }
    val buttonPx = with(density) { MIC_CORNER_BUTTON.toPx() }
    val haptic = LocalHapticFeedback.current
    val scope = rememberCoroutineScope()

    var dragging by remember { mutableStateOf(false) }
    val sharedMode = when (talkMode) {
        TalkMode.Idle -> MicCornerMode.IDLE
        TalkMode.Hold -> MicCornerMode.HOLD
        TalkMode.Continuous -> MicCornerMode.LOCKED
    }
    // The optimistic preview is presentation-only while the shared FSM processes press.
    val renderedMode = if (dragging && sharedMode == MicCornerMode.IDLE) {
        MicCornerMode.HOLD
    } else sharedMode
    val drag = remember { Animatable(0f) }
    val armed by remember(travelPx) { derivedStateOf { isArmed(drag.value, travelPx) } }
    val railShown = dragging || renderedMode == MicCornerMode.LOCKED

    val currentOnPress by rememberUpdatedState(onPress)
    val currentOnRelease by rememberUpdatedState(onRelease)
    val currentOnLock by rememberUpdatedState(onLock)
    val currentOnStopContinuous by rememberUpdatedState(onStopContinuous)
    val currentEnsurePermission by rememberUpdatedState(ensureMicPermission)

    // Shared Idle is the only teardown presentation reset; it emits no intent.
    LaunchedEffect(talkMode, dragging) {
        if (talkMode == TalkMode.Idle && !dragging) {
            scope.launch { drag.animateTo(0f, micCornerSpring()) }
        }
    }

    // Mode transition → ONE SDK talk-mode intent. Pure gesture→intent translation with zero
    // mode semantics (the SDK's TalkModeController owns them all): this only names which
    // intent each FSM edge maps to. HOLD→LOCKED now emits onLock (audio.end + semantic
    // audio.start) rather than staying a silent visual promotion.
    fun setMode(next: MicCornerMode, trigger: String) {
        val prev = renderedMode
        if (prev == next) return
        // TalkMode remains the sole state owner; this adapter only translates the
        // pointer outcome into one shared intent.
        log.info("gesture-outcome", mapOf("from" to prev.name, "to" to next.name, "trigger" to trigger))
        when {
            prev == MicCornerMode.IDLE && next == MicCornerMode.HOLD -> currentOnPress()
            prev == MicCornerMode.HOLD && next == MicCornerMode.IDLE -> currentOnRelease()
            prev == MicCornerMode.HOLD && next == MicCornerMode.LOCKED -> currentOnLock()
            prev == MicCornerMode.LOCKED && next == MicCornerMode.IDLE -> currentOnStopContinuous()
            // Not reachable via pointer today, but keep the FSM total: a direct
            // IDLE→LOCKED composes press+lock (Idle→Hold→Continuous).
            prev == MicCornerMode.IDLE && next == MicCornerMode.LOCKED -> { currentOnPress(); currentOnLock() }
        }
    }

    Box(
        modifier = modifier
            .size(MIC_CORNER_WRAP_WIDTH, MIC_CORNER_BUTTON)
            .testTag(if (renderedMode == MicCornerMode.LOCKED) "mic-corner-locked" else "mic-corner")
            .then(micCornerOverlays(drag = { drag.value }, travelPx = travelPx, railShown = railShown, armed = armed, locked = renderedMode == MicCornerMode.LOCKED))
            .pointerInput(travelPx) {
                awaitEachGesture {
                    val down = awaitFirstDown()
                    // Only presses landing on the button start a gesture — the
                    // empty rail area to its left stays inert.
                    val buttonLeft = size.width - buttonPx - drag.value
                    if (down.position.x < buttonLeft || down.position.x > buttonLeft + buttonPx) return@awaitEachGesture
                    if (renderedMode == MicCornerMode.IDLE && !currentEnsurePermission()) return@awaitEachGesture
                    down.consume()

                    val origin = renderedMode
                    val base = if (origin == MicCornerMode.LOCKED) travelPx else 0f
                    val startX = down.position.x
                    var dragPx = base
                    dragging = true
                    scope.launch { drag.snapTo(base) }
                    if (origin == MicCornerMode.IDLE) setMode(MicCornerMode.HOLD, "pointer-down")

                    // Track the finger raw (snap, no animation) until lift/cancel.
                    while (true) {
                        val event = awaitPointerEvent()
                        val change = event.changes.firstOrNull { it.id == down.id } ?: break
                        if (!change.pressed) {
                            change.consume()
                            break
                        }
                        change.consume()
                        dragPx = clampDrag(base, startX, change.position.x, travelPx)
                        scope.launch { drag.snapTo(dragPx) }
                    }

                    val outcome = resolveRelease(origin, dragPx, travelPx)
                    log.debug(
                        "release",
                        mapOf(
                            "origin" to origin.name,
                            "drag" to dragPx.roundToInt(),
                            "travel" to travelPx.roundToInt(),
                            "outcome" to outcome.mode.name,
                        ),
                    )
                    dragging = false
                    if (outcome.mode == MicCornerMode.LOCKED && origin == MicCornerMode.IDLE) {
                        haptic.performHapticFeedback(HapticFeedbackType.LongPress)
                    }
                    if (outcome.mode == MicCornerMode.IDLE && origin == MicCornerMode.LOCKED) {
                        haptic.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                    }
                    scope.launch { drag.animateTo(outcome.drag, micCornerSpring()) }
                    setMode(outcome.mode, "pointer-up")
                }
            },
    ) {
        MicCornerButton(
            mode = renderedMode,
            railShown = railShown,
            armed = armed,
            dragOffset = { -drag.value },
            modifier = Modifier.align(Alignment.CenterEnd),
        )
    }
}

File: ios/App/Chat/composer/MicCorner.swift
// ---------------------------------------------------------------------------
// MicCorner — hold-to-talk / drag-to-lock mic control welded onto the
// composer card's top-right edge (half overhanging). Combined push-to-talk +
// toggle-to-talk, mirroring the webui control
// (gateway/webui/src/components/dock/mic-corner.tsx):
//
//   press → hold (mic on immediately) · drag LEFT along the travel (button
//   tracks the finger raw, clamped 0…travel) · release ≥40% travel → locked
//   (snap to far end, haptic) · release <40% → idle (spring back, mic off).
//   From locked: press again (base = travel), drag back right; release ≤50%
//   travel → idle (mic off, haptic); else snap back locked.
//
// The pure FSM lives in MicCornerGesture.swift; visual layers + constants in
// MicCornerVisuals.swift. The control owns only its gesture state — each FSM
// transition is translated to ONE SDK talk-mode intent (pure gesture→intent, zero
// mode semantics: the SDK's TalkModeController owns them all) and the composer's
// takeover follows onModeChange. `micActive` is the external-sync input: a true→false
// edge while held/locked and not dragging (disconnect, teardown, or an SDK
// start that failed downstream) resets the control to idle WITHOUT emitting an
// intent.
//
// beginPress is the permission gate (Composer+Mic.onMicPressGate): a press
// from idle only enters hold when it returns true; a press from locked never
// re-asks (the mic is already running).
//
// accessibilityIdentifier: chat-mic (switch-like; value reflects locked;
// VoiceOver activate toggles the hands-free lock — webui keyboard parity).
// ---------------------------------------------------------------------------
import SwiftUI
import UIKit
import MobileData

struct MicCorner: View {
    /// Mirrors voiceMode == .active — external sync input, never written here.
    let talkMode: TalkMode
    /// Permission gate for a press from idle: true → enter hold + start.
    let beginPress: () -> Bool
    /// Composer takeover (waveform, hidden buttons) follows the reported mode.
    /// idle→hold (press): enter push-to-talk.
    let onPress: () -> Void
    /// hold→idle (release below the lock threshold): end the manual turn.
    let onRelease: () -> Void
    /// hold→locked (release past the lock threshold / slide-to-lock): enter continuous.
    let onLock: () -> Void
    /// locked→idle (unlock release / tap-to-stop): leave continuous.
    let onStopContinuous: () -> Void

    private var mode: MicCornerMode {
        switch talkMode {
        case .idle: return .idle
        case .hold: return .hold
        case .continuous: return .locked
        }
    }
    /// Button offset toward the lock end, 0…travel. Raw while dragging;
    /// spring-animated on release/reset.
    @State private var drag: CGFloat = 0
    @State private var dragging = false
    /// Press denied by the permission gate — swallow the rest of this gesture.
    @State private var pressRejected = false
    /// Mode the active gesture started from (resolves the release).
    @State private var origin: MicCornerMode = .idle
    /// Drag offset the gesture started from (0 from idle, travel from locked).
    @State private var base: CGFloat = 0
    /// Increments per hold entry — re-keys the ripple so it replays.
    @State private var rippleTick = 0
    @State private var glyphScale: CGFloat = 1
    @State private var bodyScale: CGFloat = 1
    @State private var glyphDimmed = false
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    private let log = AppLog("composer", "mic-corner")

    // ── Derived ───────────────────────────────────────────────────────────

    /// Live = ember styling (accent glyph / border / glow): dragging or locked.
    private var isLive: Bool { dragging || mode == .locked }
    private var isArmed: Bool { MicCornerGesture.isArmed(drag: drag, travel: MicCornerLayout.travel) }
    private var progress: CGFloat { min(1, max(0, drag / MicCornerLayout.travel)) }
    private var breatheActive: Bool { mode == .idle && !dragging && !reduceMotion }

    // ── Body ──────────────────────────────────────────────────────────────

    var body: some View {
        ZStack {
            MicCornerTrail(progress: progress)
                .frame(maxWidth: .infinity, alignment: .trailing)
                .padding(.trailing, MicCornerLayout.buttonSize / 2)
                .allowsHitTesting(false)
            MicCornerDetent(visible: isLive && mode != .locked, armed: isArmed)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.leading, MicCornerLayout.detentInset)
                .allowsHitTesting(false)
            button
                .frame(maxWidth: .infinity, alignment: .trailing)
        }
        .frame(width: MicCornerLayout.wrapWidth, height: MicCornerLayout.buttonSize)
        .onChange(of: talkMode) { _, now in
            // Shared Idle is authoritative after teardown/failure; only reset the
            // presentation offset and never emit another stop intent.
            if now == .idle, !dragging {
                withAnimation(MicCornerMotion.settle) { drag = 0 }
            }
        }
        .task(id: breatheActive) {
            // Idle rest hint — glyph breathes opacity .85↔1 (webui
            // mic-corner-breathe); suppressed under Reduce Motion.
            if breatheActive {
                withAnimation(
                    .easeInOut(duration: MicCornerMotion.breatheHalfPeriod).repeatForever(autoreverses: true)
                ) { glyphDimmed = true }
            } else {
                withAnimation(.easeOut(duration: 0.15)) { glyphDimmed = false }
            }
        }
    }

    // ── Button ────────────────────────────────────────────────────────────

    private var button: some View {
        ZStack {
            MicCornerBody(live: isLive, locked: mode == .locked, armed: isArmed, scale: bodyScale)
            if rippleTick > 0, !reduceMotion {
                MicCornerRipple().id(rippleTick)
            }
            Image(systemName: "mic")
                .font(.system(size: ComposerLayout.glyph))
                .foregroundStyle(isLive ? DuskColors.accent : DuskColors.ink3)
                .opacity(glyphDimmed ? MicCornerMotion.breatheDimOpacity : 1)
                .scaleEffect(glyphScale)
        }
        .frame(width: MicCornerLayout.buttonSize, height: MicCornerLayout.buttonSize)
        .contentShape(RoundedRectangle(cornerRadius: MicCornerLayout.cornerRadius))
        .offset(x: -drag)
        .highPriorityGesture(dragGesture)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(
            mode == .locked
                ? "Hands-free listening on — drag back to stop"
                : "Hold to talk; slide to lock hands-free"
        )
        .accessibilityValue(mode == .locked ? "locked" : "off")
        .accessibilityAddTraits(.isButton)
        .accessibilityIdentifier("chat-mic")
        .accessibilityAction { toggleLockForAccessibility() }
    }

    // ── Gesture ───────────────────────────────────────────────────────────

    private var dragGesture: some Gesture {
        DragGesture(minimumDistance: 0)
            .onChanged { value in handleDragChanged(value) }
            .onEnded { _ in handleDragEnded() }
    }

    private func handleDragChanged(_ value: DragGesture.Value) {
        if !dragging, !pressRejected {
            beginGesture()
        }
        guard dragging else { return }
        // Leftward finger = startX > currentX → positive drag (webui clampDrag
        // math). Raw tracking — no animation while the finger is down.
        drag = MicCornerGesture.clampDrag(
            base: base,
            startX: value.startLocation.x,
            currentX: value.location.x,
            travel: MicCornerLayout.travel
        )
    }

    private func beginGesture() {
        if mode == .locked {
            origin = .locked
            base = MicCornerLayout.travel
        } else {
            guard beginPress() else {
                // No record permission (denied notice shown or request just
                // fired) — do NOT enter hold; ignore the rest of this gesture.
                pressRejected = true
                log.info("press-rejected reason=permission")
                return
            }
            origin = .idle
            base = 0
            fireRipple()
            setMode(.hold, trigger: "press")
        }
        drag = base
        dragging = true
    }

    private func handleDragEnded() {
        pressRejected = false
        guard dragging else { return }
        dragging = false
        let travel = MicCornerLayout.travel
        let outcome = MicCornerGesture.resolveRelease(origin: origin, drag: drag, travel: travel)
        log.info(
            "release origin=\(origin.rawValue) drag=\(Int(drag)) travel=\(Int(travel)) outcome=\(outcome.mode.rawValue)"
        )
        withAnimation(MicCornerMotion.settle) { drag = outcome.drag }
        if outcome.mode == .locked, origin == .idle { impact(.medium) }
        if outcome.mode == .idle, origin == .locked { impact(.light) }
        setMode(outcome.mode, trigger: "release")
    }

    // ── Mode transitions ──────────────────────────────────────────────────

    private func setMode(_ next: MicCornerMode, trigger: String) {
        guard next != mode else { return }
        let prev = mode
        log.info("mode-change from=\(prev.rawValue) to=\(next.rawValue) trigger=\(trigger)")
        if next == .locked { snapBounce() }
        emitIntent(from: prev, to: next)
    }

    /// Pure gesture→intent translation — the one place the corner speaks to the SDK.
    /// Zero mode semantics: it only names which talk-mode intent each FSM transition maps
    /// to; the TalkModeController owns interrupt-on-press, turnMode, and buffer-and-defer.
    /// idle→locked (VoiceOver direct-lock) composes press+lock (Idle→Hold→Continuous).
    private func emitIntent(from: MicCornerMode, to: MicCornerMode) {
        switch (from, to) {
        case (.idle, .hold): onPress()
        case (.hold, .idle): onRelease()
        case (.hold, .locked): onLock()
        case (.locked, .idle): onStopContinuous()
        case (.idle, .locked): onPress(); onLock()
        default: break
        }
    }

    /// VoiceOver can't drag — activate toggles the hands-free lock directly
    /// (parity with the webui keyboard path).
    private func toggleLockForAccessibility() {
        if mode == .locked {
            withAnimation(MicCornerMotion.settle) { drag = 0 }
            impact(.light)
            setMode(.idle, trigger: "accessibility")
        } else if beginPress() {
            withAnimation(MicCornerMotion.settle) { drag = MicCornerLayout.travel }
            impact(.medium)
            setMode(.locked, trigger: "accessibility")
        }
    }

    // ── Motion + haptics ──────────────────────────────────────────────────

    private func fireRipple() {
        guard !reduceMotion else { return }
        rippleTick += 1
    }

    /// Lock snap — glyph + body spring back from an undershoot so they
    /// overshoot past 1 and settle (webui mic-snap / mic-settle keyframes).
    private func snapBounce() {
        guard !reduceMotion else { return }
        glyphScale = MicCornerMotion.snapGlyphStartScale
        bodyScale = MicCornerMotion.snapBodyStartScale
        withAnimation(MicCornerMotion.snap) {
            glyphScale = 1
            bodyScale = 1
        }
    }

    private func impact(_ style: UIImpactFeedbackGenerator.FeedbackStyle) {
        UIImpactFeedbackGenerator(style: style).impactOccurred()
    }
}

#Preview("Corner mic — idle (interactive)") {
    MicCorner(
        talkMode: .idle,
        beginPress: { true },
        onPress: {},
        onRelease: {},
        onLock: {},
        onStopContinuous: {}
    )
    .padding(Space.xxl)
    .background(DuskColors.paper)
    .padding(Space.xxl)
    .background(DuskColors.bg)
}
